x/src/lint/engine/mod.rs

```diff
@@ -0,0 +1,207 @@
+    NonUtf8,
+    /// This extension was unsupported.
+    UnsupportedExtension(Option<&'l str>),
+    // TODO: Add more reasons.
+}
+
+/// A message raised by a lint.
+#[derive(Debug)]
+pub struct LintMessage {
+    level: LintLevel,
+    message: Cow<'static, str>,
+}
+
+impl LintMessage {
+    pub fn new(level: LintLevel, message: impl Into<Cow<'static, str>>) -> Self {
+        Self {
+            level,
+            message: message.into(),
+        }
+    }
+
+    pub fn level(&self) -> LintLevel {
+        self.level
+    }
+
+    pub fn message(&self) -> &str {
+        &self.message
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+#[allow(dead_code)]
+#[non_exhaustive]
+pub enum LintLevel {
+    Error,
+    Warning,
+    // TODO: add more levels?
+}
+
+impl fmt::Display for LintLevel {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            LintLevel::Error => write!(f, "ERROR"),
+            LintLevel::Warning => write!(f, "WARNING"),
+        }
+    }
+}
+
+/// Message source for lints.
+#[derive(Copy, Clone, Debug)]
+pub struct LintSource<'l> {
+    name: &'static str,
+    kind: LintKind<'l>,
+}
+
+impl<'l> LintSource<'l> {
+    pub(super) fn new(name: &'static str, kind: LintKind<'l>) -> Self {
+        Self { name, kind }
+    }
+
+    pub fn name(&self) -> &'static str {
+        self.name
+    }
+
+    pub fn kind(&self) -> LintKind<'l> {
+        self.kind
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum LintKind<'l> {
+    Project,
+    File(&'l Path),
+    Content(&'l Path),
+}
+
+impl<'l> fmt::Display for LintKind<'l> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            LintKind::Project => write!(f, "project"),
+            LintKind::File(path) => write!(f, "file {}", path.display()),
+            LintKind::Content(path) => write!(f, "content {}", path.display()),
+        }
+    }
+}
+
+/// Type alias for the return type for `run` methods.
+pub type Result<T> = result::Result<T, SystemError>;
+
+/// A system error that happened while running a lint.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum SystemError {
+    Exec {
+        cmd: &'static str,
+        status: ExitStatus,
+    },
+    Io(io::Error),
+}
+
+impl fmt::Display for SystemError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            SystemError::Exec { cmd, status } => match status.code() {
+                Some(code) => write!(f, "'{}' failed with exit code {}", cmd, code),
+                None => write!(f, "'{}' terminated by signal", cmd),
+            },
+            SystemError::Io(err) => write!(f, "IO error: {}", err),
+        }
+    }
+}
+
+impl error::Error for SystemError {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match self {
+            SystemError::Exec { .. } => None,
+            SystemError::Io(err) => Some(err),
+        }
+    }
+}
```

Comment:
Not that this matters but you could always use `thiserror` if you want to derive the display and error traits, I tend to like it because it makes adding From impls fairly easy.

Reply:
Hmm, I'm worried about the perf cost of adding `syn` and `quote` to the dependency graph of a devtool like this... I can switch if you think it's worth it though.

Reply:
No its not particularly important to switch to atm, especially given you've already written this up. We can reevaluate later

Reply:
No its not particularly important to switch to atm, especially given you've already written this up. We can reevaluate later. (though atm i think x depends on syn and quote)

Reply:
Ah if it already does it might make more sense